source/common/router/upstream_request.cc

```diff
@@ -94,13 +94,15 @@ UpstreamRequest::UpstreamRequest(RouterFilterInterface& parent,
           Runtime::runtimeFeatureEnabled("envoy.reloadable_features.allow_upstream_filters")),
       stream_options_({can_send_early_data, can_use_http3}) {
   if (parent_.config().start_child_span_) {
-    span_ = parent_.callbacks()->activeSpan().spawnChild(
-        parent_.callbacks()->tracingConfig(),
-        absl::StrCat("router ", parent.cluster()->observabilityName(), " egress"),
-        parent.timeSource().systemTime());
-    if (parent.attemptCount() != 1) {
-      // This is a retry request, add this metadata to span.
-      span_->setTag(Tracing::Tags::get().RetryCount, std::to_string(parent.attemptCount() - 1));
+    if (auto tracing_config = parent_.callbacks()->tracingConfig(); tracing_config.has_value()) {
```

Comment:
If there's no HCM trace config. The `activeSpan()` will return a `NullSpanImpl` instance. All these spans and child spans trace nothing. So, I don't think it's a function change.
I'd think this PR is more like a minor optimization to your previous fix.

Reply:
hmmm, I had assumed the folks who did the prior bug report had tracing working.  Would it be possible to add actual tracing checks to the RouterOnlyTracing to verify there's no behavioral change with this PR?

Reply:
🤔 If we can ensure the `parent_.callbacks()->activeSpan().spawnChild()` call will return a `NullSpanImpl` instance when there is no HCM tracing config, then we can ensure there's no behavior change for the router only tracinig. Because even the child span is created when the `start_child_span` is set to `true`, it's a null span which do nothing.
May be I can add a test to guard it.